### PR TITLE
Tickets/dm-2177

### PR DESCRIPTION
**Pull request by wangd**

### CODE REVIEW

**Comment by ktlim** [comment]
Is this too hard to write test cases for?

File: core/modules/czar/python/appInterface.py [hunk 1]
```diff
@@ -30,6 +32,17 @@
...
+def parseKillId(killQuery):
+    """From a "KILL QUERY 1234" string, return 1234"""
+    tokens = map(string.strip, killQuery.split())
+    try:
+        # Could add validation code here, but shouldn't be necessary.
+        return int(tokens[2])
+    except:
+        return None
+
+
```
Comment (jgates108) [inline core/modules/czar/python/appInterface.py]: Extra blank line.

File: core/modules/czar/python/appInterface.py [hunk 2]
```diff
@@ -101,14 +113,39 @@ def submitQuery(self, query, conditions):
...
+        proxyName = conditions["client_dst_name"]
+        proxyThread = conditions["server_thread_id"]
+        self._clientToServerId[(proxyThread, proxyName)] = a.sessionId
+
         return (resultName, lockName, "")
 
-    def killQuery(self, query, taskId):
+    def killQuery(self, sessionId):
         """Process a kill query command (experimental).  Params: query, taskId."""
         ## Disable query killing DM-1715
```
Comment (jgates108) [inline core/modules/czar/python/appInterface.py]: Shouldn't this comment also be deleted?

File: core/modules/czar/python/appInterface.py [hunk 3]
```diff
@@ -101,14 +113,39 @@ def submitQuery(self, query, conditions):
...
-        ##a = app.KillQueryAction(query)
-        ##self._callWithThread(a.invoke)
-        return "Attempt query kill: " + query
+        a = app.KillQueryAction(sessionId)
+        self._maybeCallWithThread(a.invoke)
+        return "Attempt query kill: " + str(sessionId)
+
+    def killQueryUgly(self, killStr, clientId):
+        """Process a kill query command (experimental).  Params: killquery, clientName."""
+        #lookup sessionId using thread_id (parsed from killStr) and clientName
```
Comment (jgates108) [inline core/modules/czar/python/appInterface.py]: I think killquery should probably be killQuery or killStr and clientNames should probably be clientId.

File: core/modules/czar/python/appInterface.py [hunk 4]
```diff
@@ -101,14 +113,39 @@ def submitQuery(self, query, conditions):
...
+        # Remember client context for kill-operations
+        proxyName = conditions["client_dst_name"]
+        proxyThread = conditions["server_thread_id"]
+        self._clientToServerId[(proxyThread, proxyName)] = a.sessionId
+
         return (resultName, lockName, "")
 
-    def killQuery(self, query, taskId):
+    def killQuery(self, sessionId):
         """Process a kill query command (experimental).  Params: query, taskId."""
```
Comment (jgates108) [inline core/modules/czar/python/appInterface.py]: query, taskId should be replaced by sessionId. I'm of the opinion that repeating function arguments in comments without explanations is not useful, and then it's only useful if the meaning isn't clear from the name or there's other important information, like units.

File: core/modules/wbase/Task.cc [hunk 5]
```diff
@@ -147,14 +147,21 @@ void Task::poison() {
...
+            }
         }
     }
-    _poisonFunc = poisonFunc;
+    if(shouldCall) {
+        (*poisonFunc)();
+    } else {
+        _poisonFunc = poisonFunc;
+    }
 }
```
Comment (jgates108) [inline core/modules/wbase/Task.cc]: (_poisonFunc)() is called in Task::poison() with _mutex locked, and (*poisonFunc)() is called here without _mutex locked. This should probably be done consistently one way or the other.

Comment (wangd) [inline core/modules/wbase/Task.cc]: fair enough.

**Pull request merged**